#   Gurgaon Data Preprocessing(Flats)

## 1.1 Introduction and Setup
In this phase we will first import the raw and scraped data for the gurgaon flats and perform an analysis. To analyse we will first import required libraries and load our dataset in the pandas dataframe to know what we are working with.

In [45]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [46]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [47]:
df = pd.read_excel('C:\\capstone_project_dscience\\real_estate_project\\dataset\\flats.xlsx')

## 1.2 High Level Data Inspection
here we will see what are the dimensions of our dataset, how many attributes are there, attributes name and spot any missing values

In [48]:
df.head()

,property_name,link,society,price,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,address,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating,property_id
0,2 BHK Flat in Krishna Colony,https://www.99acres.com/2-bhk-bedroom-apartmen...,maa bhagwati residency,45 Lac,"₹ 5,000/sq.ft.",Carpet area: 900 (83.61 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"Krishna Colony, Gurgaon, Haryana",4th of 4 Floors,West,1 to 5 Year Old,"['Chintapurni Mandir', 'State bank ATM', 'Pear...",So with lift.Maa bhagwati residency is one of ...,"['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...","['Feng Shui / Vaastu Compliant', 'Security / F...","['Environment4 out of 5', 'Safety4 out of 5', ...",C68850746
1,2 BHK Flat in Ashok Vihar,https://www.99acres.com/2-bhk-bedroom-apartmen...,Apna Enclave,50 Lac,"₹ 7,692/sq.ft.",Carpet area: 650 (60.39 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"46b, Ashok Vihar, Gurgaon, Haryana",1st of 3 Floors,West,10+ Year Old,"['Chintapurni Mandir', 'Sheetla Mata Mandir', ...","Property situated on main road, railway statio...","['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...","['Security / Fire Alarm', 'Maintenance Staff',...","['Environment4 out of 5', 'Safety4 out of 5', ...",H68850564
2,2 BHK Flat in Sohna,https://www.99acres.com/2-bhk-bedroom-apartmen...,Tulsiani Easy in Homes,40 Lac,"₹ 6,722/sq.ft.",Carpet area: 595 (55.28 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,NaN,"Sohna, Gurgaon, Haryana",12nd of 14 Floors,NaN,0 to 1 Year Old,"['Huda City Metro', 'Golf Course extn road', '...","This property is 15 km away from badshapur, gu...",NaN,"['Power Back-up', 'Feng Shui / Vaastu Complian...","['Environment4 out of 5', 'Safety4 out of 5', ...",J68850120
3,2 BHK Flat in Sector 61 Gurgaon,https://www.99acres.com/2-bhk-bedroom-apartmen...,Smart World Orchard,1.47 Crore,"₹ 12,250/sq.ft.",Carpet area: 1200 (111.48 sq.m.),2 Bedrooms,2 Bathrooms,2 Balconies,Study Room,"Sector 61 Gurgaon, Gurgaon, Haryana",2nd of 4 Floors,NaN,2023-12-01 00:00:00,"['Sector 55-56 Metro station', 'Bestech Centra...",Near to metro station of sector 56 and opposit...,NaN,"['Security / Fire Alarm', 'Private Garden / Te...",NaN,S68849476
4,2 BHK Flat in Sector 92 Gurgaon,https://www.99acres.com/2-bhk-bedroom-apartmen...,Parkwood Westend,70 Lac,"₹ 5,204/sq.ft.",Super Built up area 1345(124.95 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,Study Room,"Sector 92 Gurgaon, Gurgaon, Haryana",5th of 8 Floors,NaN,Under Construction,"['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...",We are the proud owners of this 2 bhk alongwit...,[],NaN,"['Environment5 out of 5', 'Safety3 out of 5', ...",L47956793


In [49]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3028 entries, 0 to 3027
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype 
---  ------           --------------  ----- 
 0   property_name    3028 non-null   str   
 1   link             3028 non-null   str   
 2   society          3027 non-null   str   
 3   price            3008 non-null   str   
 4   area             3015 non-null   str   
 5   areaWithType     3009 non-null   str   
 6   bedRoom          3009 non-null   str   
 7   bathroom         3009 non-null   str   
 8   balcony          3009 non-null   str   
 9   additionalRoom   1695 non-null   str   
 10  address          3003 non-null   str   
 11  floorNum         3007 non-null   str   
 12  facing           2128 non-null   str   
 13  agePossession    3008 non-null   object
 14  nearbyLocations  2914 non-null   str   
 15  description      3009 non-null   str   
 16  furnishDetails   2204 non-null   str   
 17  features         2595 non-null   str   
 18 

### 1.2.1 column dtype
        1) columns like price and area they are also of string type instead of numerical values
        2) columns like 'link' and 'property_id' are of no use

In [50]:
df.drop(columns=['link','property_id'], inplace=True)

In [51]:
df.shape

(3028, 18)

### 1.2.2 dataset shape
        1)number of rows in the dataset is 3028
        2) number of columns -> 20

## 1.3 cleaning and feature extraction
        property_name column
            1) i want to extract the sector name from the property_name column


In [52]:
df['property_name'].value_counts()

property_name
2 BHK Flat in Sohna                                126
3 BHK Flat in Sector 102 Gurgaon                    76
3 BHK Flat in Sector 65 Gurgaon                     57
3 BHK Flat in Sector 92 Gurgaon                     53
3 BHK Flat in Sector 69 Gurgaon                     51
2 BHK Flat in Sector 85 Gurgaon                     49
3 BHK Flat in Sector 83 Gurgaon                     47
3 BHK Flat in Sector 81 Gurgaon                     45
3 BHK Flat in Sector 79 Gurgaon                     44
3 BHK Flat in Sector 109 Gurgaon                    44
2 BHK Flat in Sector 107 Gurgaon                    39
3 BHK Flat in Sector 37D Gurgaon                    38
3 BHK Flat in Sector 90 Gurgaon                     38
2 BHK Flat in Sector 79 Gurgaon                     37
3 BHK Flat in Sector 85 Gurgaon                     36
3 BHK Flat in Sector 70 Gurgaon                     36
3 BHK Flat in Sector 37C Gurgaon                    35
3 BHK Flat in Sector 86 Gurgaon                    

     here i extracted the sector detail from the property_name column

In [53]:
sectors = df['property_name'].str.split('in').str.get(1).str.split('Gurgaon').str.get(0).str.strip()

    created a new column for the sectors

In [54]:
df.insert(1, 'sector', sectors)

In [55]:
df.head()

,property_name,sector,society,price,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,address,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating
0,2 BHK Flat in Krishna Colony,Krishna Colony,maa bhagwati residency,45 Lac,"₹ 5,000/sq.ft.",Carpet area: 900 (83.61 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"Krishna Colony, Gurgaon, Haryana",4th of 4 Floors,West,1 to 5 Year Old,"['Chintapurni Mandir', 'State bank ATM', 'Pear...",So with lift.Maa bhagwati residency is one of ...,"['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...","['Feng Shui / Vaastu Compliant', 'Security / F...","['Environment4 out of 5', 'Safety4 out of 5', ..."
1,2 BHK Flat in Ashok Vihar,Ashok Vihar,Apna Enclave,50 Lac,"₹ 7,692/sq.ft.",Carpet area: 650 (60.39 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"46b, Ashok Vihar, Gurgaon, Haryana",1st of 3 Floors,West,10+ Year Old,"['Chintapurni Mandir', 'Sheetla Mata Mandir', ...","Property situated on main road, railway statio...","['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...","['Security / Fire Alarm', 'Maintenance Staff',...","['Environment4 out of 5', 'Safety4 out of 5', ..."
2,2 BHK Flat in Sohna,Sohna,Tulsiani Easy in Homes,40 Lac,"₹ 6,722/sq.ft.",Carpet area: 595 (55.28 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,NaN,"Sohna, Gurgaon, Haryana",12nd of 14 Floors,NaN,0 to 1 Year Old,"['Huda City Metro', 'Golf Course extn road', '...","This property is 15 km away from badshapur, gu...",NaN,"['Power Back-up', 'Feng Shui / Vaastu Complian...","['Environment4 out of 5', 'Safety4 out of 5', ..."
3,2 BHK Flat in Sector 61 Gurgaon,Sector 61,Smart World Orchard,1.47 Crore,"₹ 12,250/sq.ft.",Carpet area: 1200 (111.48 sq.m.),2 Bedrooms,2 Bathrooms,2 Balconies,Study Room,"Sector 61 Gurgaon, Gurgaon, Haryana",2nd of 4 Floors,NaN,2023-12-01 00:00:00,"['Sector 55-56 Metro station', 'Bestech Centra...",Near to metro station of sector 56 and opposit...,NaN,"['Security / Fire Alarm', 'Private Garden / Te...",NaN
4,2 BHK Flat in Sector 92 Gurgaon,Sector 92,Parkwood Westend,70 Lac,"₹ 5,204/sq.ft.",Super Built up area 1345(124.95 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,Study Room,"Sector 92 Gurgaon, Gurgaon, Haryana",5th of 8 Floors,NaN,Under Construction,"['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...",We are the proud owners of this 2 bhk alongwit...,[],NaN,"['Environment5 out of 5', 'Safety3 out of 5', ..."


In [56]:
df['sector'] = df['sector'].str.lower()

In [57]:
df['sector'].value_counts()

sector
sohna                            160
sector 102                       113
sector 85                        110
sector 92                        100
sector 69                         92
sector 90                         90
sector 65                         89
sector 81                         87
sector 79                         82
sector 83                         68
sector 104                        66
sector 37d                        66
sector 86                         63
sector 107                        61
sector 95                         58
sector 108                        55
sector 89                         55
sector 37c                        55
sector 70                         53
sector 109                        52
sector 84                         51
sector-33 sohna                   49
sector 56                         44
sector 113                        43
sector 67                         42
sector 61                         41
sector 103                     

***society column** <br>
        here in society column we can see that the some digits as well as characters are also attached with the names

In [58]:
df['society'].value_counts()

society
SS The Leaf3.8 ★                                     73
Tulip Violet4.3 ★                                    40
Shapoorji Pallonji Joyville Gurugram4.0 ★            39
Signature Global Park4.0 ★                           36
Shree Vardhman Victoria3.8 ★                         35
Tulip Violet4.2 ★                                    33
Smart World Orchard                                  32
Smart World Gems                                     32
Emaar MGF Emerald Floors Premier3.8 ★                32
Paras Dews                                           31
DLF The Ultima4.0 ★                                  31
DLF Regal Gardens3.9 ★                               30
Shree Vardhman Flora3.8 ★                            29
M3M Woodshire4.0 ★                                   29
La Vida by Tata Housing                              28
Signature Global Solera3.7 ★                         28
Godrej Nature Plus                                   27
BPTP Terra3.8 ★                         

In [59]:
df['society'] = df['society'].str.lower()


    now lets remove these digits and characters from the names

In [60]:
df['society'] = df['society'].str.replace(r'[\d\★]', '', regex=True)

In [61]:
df['society'].value_counts()

society
ss the leaf.                                      73
tulip violet.                                     73
shapoorji pallonji joyville gurugram.             39
signature global park.                            36
shree vardhman victoria.                          35
dlf new town heights .                            34
smart world orchard                               32
smart world gems                                  32
emaar mgf emerald floors premier.                 32
paras dews                                        31
dlf the ultima.                                   31
dlf regal gardens.                                30
shree vardhman flora.                             29
mm woodshire.                                     29
la vida by tata housing                           28
signature global solera.                          28
godrej nature plus                                27
bptp terra.                                       25
emaar gurgaon greens.                 

**price**<br>


In [62]:
df['price'].value_counts()

price
1.25 Crore          79
1.1 Crore           61
1.4 Crore           60
1.2 Crore           59
1.5 Crore           59
90 Lac              58
1.3 Crore           57
95 Lac              53
2 Crore             51
1.75 Crore          47
1 Crore             46
1.6 Crore           43
1.35 Crore          41
1.55 Crore          40
1.9 Crore           40
75 Lac              38
1.65 Crore          38
1.8 Crore           37
1.7 Crore           37
80 Lac              36
2.2 Crore           34
50 Lac              33
1.15 Crore          33
85 Lac              31
1.45 Crore          31
1.05 Crore          30
40 Lac              29
2.5 Crore           29
60 Lac              29
45 Lac              26
2.1 Crore           26
65 Lac              25
1.85 Crore          23
2.35 Crore          23
35 Lac              23
3 Crore             22
70 Lac              21
2.25 Crore          20
55 Lac              20
3.5 Crore           19
2.4 Crore           18
2.3 Crore           18
30 Lac              17
2.65 

In [63]:
df['price'] = df['price'].str.lower()

    the price column is string dtype here. so i converted every value(except price on value and NAN) to crore and make them float type

In [64]:
def price_to_number(price):
    if pd.isna(price):
        return np.nan
    if 'lac' in price:
        return float(price.replace('lac','').strip())/100
    elif 'crore' in price:
        return float(price.replace('crore','').strip())
    else: return price
    

df['price'] = df['price'].apply(price_to_number)

In [65]:
df.head()

,property_name,sector,society,price,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,address,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating
0,2 BHK Flat in Krishna Colony,krishna colony,maa bhagwati residency,0.45,"₹ 5,000/sq.ft.",Carpet area: 900 (83.61 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"Krishna Colony, Gurgaon, Haryana",4th of 4 Floors,West,1 to 5 Year Old,"['Chintapurni Mandir', 'State bank ATM', 'Pear...",So with lift.Maa bhagwati residency is one of ...,"['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...","['Feng Shui / Vaastu Compliant', 'Security / F...","['Environment4 out of 5', 'Safety4 out of 5', ..."
1,2 BHK Flat in Ashok Vihar,ashok vihar,apna enclave,0.5,"₹ 7,692/sq.ft.",Carpet area: 650 (60.39 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"46b, Ashok Vihar, Gurgaon, Haryana",1st of 3 Floors,West,10+ Year Old,"['Chintapurni Mandir', 'Sheetla Mata Mandir', ...","Property situated on main road, railway statio...","['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...","['Security / Fire Alarm', 'Maintenance Staff',...","['Environment4 out of 5', 'Safety4 out of 5', ..."
2,2 BHK Flat in Sohna,sohna,tulsiani easy in homes,0.4,"₹ 6,722/sq.ft.",Carpet area: 595 (55.28 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,NaN,"Sohna, Gurgaon, Haryana",12nd of 14 Floors,NaN,0 to 1 Year Old,"['Huda City Metro', 'Golf Course extn road', '...","This property is 15 km away from badshapur, gu...",NaN,"['Power Back-up', 'Feng Shui / Vaastu Complian...","['Environment4 out of 5', 'Safety4 out of 5', ..."
3,2 BHK Flat in Sector 61 Gurgaon,sector 61,smart world orchard,1.47,"₹ 12,250/sq.ft.",Carpet area: 1200 (111.48 sq.m.),2 Bedrooms,2 Bathrooms,2 Balconies,Study Room,"Sector 61 Gurgaon, Gurgaon, Haryana",2nd of 4 Floors,NaN,2023-12-01 00:00:00,"['Sector 55-56 Metro station', 'Bestech Centra...",Near to metro station of sector 56 and opposit...,NaN,"['Security / Fire Alarm', 'Private Garden / Te...",NaN
4,2 BHK Flat in Sector 92 Gurgaon,sector 92,parkwood westend,0.7,"₹ 5,204/sq.ft.",Super Built up area 1345(124.95 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,Study Room,"Sector 92 Gurgaon, Gurgaon, Haryana",5th of 8 Floors,NaN,Under Construction,"['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...",We are the proud owners of this 2 bhk alongwit...,[],NaN,"['Environment5 out of 5', 'Safety3 out of 5', ..."


    remove the 286 index row beacause it is of no use

In [66]:
df.iloc[285:290]

,property_name,sector,society,price,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,address,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating
285,3 BHK Flat in Sector 85 Gurgaon,sector 85,ss the leaf.,1.93,"₹ 8,075/sq.ft.",Super Built up area 2408(223.71 sq.m.)Built Up...,3 Bedrooms,4 Bathrooms,3 Balconies,Servant Room,"000000, Sector 85 Gurgaon, Gurgaon, Haryana",7th of 26 Floors,North-East,1 to 5 Year Old,"['Sapphire 83 Mall', 'Dwarka Expressway', 'Cen...",Ss the leaf is one of gurgaon's most sought af...,"['3 Wardrobe', '4 Fan', '10 Light', 'No AC', '...","['Centrally Air Conditioned', 'Water purifier'...","['Connectivity4.5 out of 5', 'Management4 out ..."
286,property_name,NaN,society,price,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,address,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating
287,4 BHK Flat in Sector 113 Gurgaon,sector 113,smart world one dxp,3.45,"₹ 13,269/sq.ft.",Super Built up area 2600(241.55 sq.m.),4 Bedrooms,4 Bathrooms,3 Balconies,"Study Room,Servant Room","A121212, Sector 113 Gurgaon , Gurgaon, Haryana",11st of 30 Floors,South-East,2025-12-01 00:00:00,"['Palam Vihar Vyapar kendra', 'Ram Mandir', 'D...",Smart world soon bringing futuristic homes of ...,"['1 Geyser', '1 Microwave', '6 Light', '6 AC',...","['Centrally Air Conditioned', 'Water purifier'...",NaN
288,3 BHK Flat in Sector 56 Gurgaon,sector 56,sukhshanti apartment,1.7,"₹ 7,727/sq.ft.",Built Up area: 2200 (204.39 sq.m.)Carpet area:...,3 Bedrooms,3 Bathrooms,3 Balconies,Study Room,"402, Sector 56 Gurgaon, Gurgaon, Haryana",4th of 8 Floors,South,5 to 10 Year Old,"['Sector metro station', 'Sector metro station...",Opposite dlf metro station\nBest for office go...,NaN,"['Feng Shui / Vaastu Compliant', 'Security / F...","['Connectivity5 out of 5', 'Lifestyle5 out of ..."
289,3 BHK Flat in Sector 65 Gurgaon,sector 65,mm golfestate.,6.15,"₹ 15,998/sq.ft.",Carpet area: 3844 (357.12 sq.m.),3 Bedrooms,5 Bathrooms,3+ Balconies,Servant Room,"Duplex 3bhk, Sector 65 Gurgaon, Gurgaon, Haryana",1st of 14 Floors,West,1 to 5 Year Old,"['Sector 54 Chowk Metro Station', 'Sahara Mall...",3bhk flat/apartment for sale in m3m golf estat...,"['1 Water Purifier', '1 Fan', '1 Fridge', '1 E...","['Feng Shui / Vaastu Compliant', 'Security / F...","['Connectivity5 out of 5', 'Management5 out of..."


In [67]:
df.drop(286, inplace=True)

    the area column name should be price per sq ft instead of area.
        1) so we will first make remvoe the sq/ft from the values and then make them numerical dtype
        2) also change the name of the column 'area' -> 'price_per_sqft'


In [68]:
df.rename(columns={'area':'price_per_sqft'},inplace=True)

In [69]:
df['price_per_sqft'] = df['price_per_sqft'].str.split('/').str.get(0).str.split(' ').str.get(1).str.replace(',','').astype(float)

    from price and price_per_sqft column i can find the area

In [70]:
area = round((df['price']/df['price_per_sqft'])*10000000,2)

In [71]:
df.insert(5, 'area', area)

In [72]:
df.head()

,property_name,sector,society,price,price_per_sqft,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,address,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating
0,2 BHK Flat in Krishna Colony,krishna colony,maa bhagwati residency,0.45,5000.0,900.0,Carpet area: 900 (83.61 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"Krishna Colony, Gurgaon, Haryana",4th of 4 Floors,West,1 to 5 Year Old,"['Chintapurni Mandir', 'State bank ATM', 'Pear...",So with lift.Maa bhagwati residency is one of ...,"['3 Fan', '4 Light', '1 Wardrobe', 'No AC', 'N...","['Feng Shui / Vaastu Compliant', 'Security / F...","['Environment4 out of 5', 'Safety4 out of 5', ..."
1,2 BHK Flat in Ashok Vihar,ashok vihar,apna enclave,0.5,7692.0,650.03,Carpet area: 650 (60.39 sq.m.),2 Bedrooms,2 Bathrooms,1 Balcony,NaN,"46b, Ashok Vihar, Gurgaon, Haryana",1st of 3 Floors,West,10+ Year Old,"['Chintapurni Mandir', 'Sheetla Mata Mandir', ...","Property situated on main road, railway statio...","['3 Wardrobe', '4 Fan', '1 Exhaust Fan', '1 Ge...","['Security / Fire Alarm', 'Maintenance Staff',...","['Environment4 out of 5', 'Safety4 out of 5', ..."
2,2 BHK Flat in Sohna,sohna,tulsiani easy in homes,0.4,6722.0,595.06,Carpet area: 595 (55.28 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,NaN,"Sohna, Gurgaon, Haryana",12nd of 14 Floors,NaN,0 to 1 Year Old,"['Huda City Metro', 'Golf Course extn road', '...","This property is 15 km away from badshapur, gu...",NaN,"['Power Back-up', 'Feng Shui / Vaastu Complian...","['Environment4 out of 5', 'Safety4 out of 5', ..."
3,2 BHK Flat in Sector 61 Gurgaon,sector 61,smart world orchard,1.47,12250.0,1200.0,Carpet area: 1200 (111.48 sq.m.),2 Bedrooms,2 Bathrooms,2 Balconies,Study Room,"Sector 61 Gurgaon, Gurgaon, Haryana",2nd of 4 Floors,NaN,2023-12-01 00:00:00,"['Sector 55-56 Metro station', 'Bestech Centra...",Near to metro station of sector 56 and opposit...,NaN,"['Security / Fire Alarm', 'Private Garden / Te...",NaN
4,2 BHK Flat in Sector 92 Gurgaon,sector 92,parkwood westend,0.7,5204.0,1345.12,Super Built up area 1345(124.95 sq.m.),2 Bedrooms,2 Bathrooms,3 Balconies,Study Room,"Sector 92 Gurgaon, Gurgaon, Haryana",5th of 8 Floors,NaN,Under Construction,"['Yadav Clinic', 'Bangali Clinic', 'Dr. J. S. ...",We are the proud owners of this 2 bhk alongwit...,[],NaN,"['Environment5 out of 5', 'Safety3 out of 5', ..."


**bedroom**<br>
    converting bedroom to numerical dtype by removing the Bedrooms string from the string

In [73]:
df['bedRoom'].value_counts()

bedRoom
3 Bedrooms    1439
2 Bedrooms     948
4 Bedrooms     481
1 Bedroom      105
5 Bedrooms      32
6 Bedrooms       3
Name: count, dtype: int64

In [74]:
df['bedRoom'] = df['bedRoom'].str.split(' ').str.get(0).str.strip().astype(float)

**bathroom**<br>
    removing the bathroom keyword from the string

In [75]:
df['bathroom'].value_counts()

bathroom
2 Bathrooms    1047
3 Bathrooms     991
4 Bathrooms     638
5 Bathrooms     170
1 Bathroom      115
6 Bathrooms      42
7 Bathrooms       5
Name: count, dtype: int64

In [76]:
df['bathroom'] = df['bathroom'].str.split(' ').str.get(0).str.strip().astype(float)

**balcony**<br>
    we can't convert it into a numerical value because there is a value '3+ Balconies'

In [77]:
df['balcony'].value_counts()

balcony
3 Balconies     977
3+ Balconies    863
2 Balconies     751
1 Balcony       318
No Balcony       99
Name: count, dtype: int64

In [78]:
df['balcony'] = df['balcony'].str.split(' ').str.get(0).str.strip()

    i am not able to get any useful information from the **address* column so i am going to remove it, but if later on if it is required then i will take the action    

In [79]:
df.drop(columns=['address'], inplace=True)

**floorNum**<br>
    1) there are values like 'Ground', 'Lower Ground', 'Basement'. so i will replace them with their respective meanings in numerical form<br>
    2) for values like 20th of 28 floors i am going to take out only the '20' value from this because this is the thing buyer cares most about(on which floor, I am getting the flat)

In [80]:
df['floorNum'].value_counts()

floorNum
2nd   of 4 Floors           74
3rd   of 4 Floors           71
4th   of 4 Floors           63
1st   of 4 Floors           61
14th   of 14 Floors         50
12nd   of 14 Floors         49
Ground of 14 Floors         40
10th   of 14 Floors         35
7th   of 14 Floors          35
8th   of 14 Floors          34
4th   of 14 Floors          28
6th   of 14 Floors          27
3rd   of 3 Floors           26
2nd   of 2 Floors           26
1st   of 14 Floors          26
3rd   of 14 Floors          24
8th   of 19 Floors          24
5th   of 14 Floors          24
11st   of 14 Floors         23
1st   of 1 Floors           23
9th   of 9 Floors           23
9th   of 14 Floors          23
2nd   of 3 Floors           22
5th   of 12 Floors          22
2nd   of 14 Floors          21
8th   of 18 Floors          20
Ground of 1 Floors          18
7th   of 18 Floors          18
6th   of 18 Floors          18
10th   of 19 Floors         18
12nd   of 12 Floors         17
10th   of 18 Floors         17

    i am creating a copy for my dataframe so that if there is any trouble i can come back to my real dataframe

In [81]:
df1 = df.copy()

In [82]:
df1['floorNum'] = df1['floorNum'].str.split('of').str.get(0).str.replace('Lower Ground', '-1').str.replace('Ground', '0').str.replace('Basement', '-1')

now lets remove the 'nd' from '2nd' etc...

In [83]:
df1['floorNum'].value_counts()

floorNum
2nd        258
3rd        249
1st        220
4th        204
8th        200
7th        189
10th       185
6th        185
5th        172
9th        171
12nd       160
0          132
14th       120
11st       115
15th        84
16th        53
17th        52
13rd        44
18th        38
19th        34
20th        25
21st        15
23rd        15
22nd        14
24th        11
25th         9
27th         7
-1           6
30th         5
26th         5
31st         4
28th         4
29th         3
33rd         3
32nd         2
43           2
35th         2
39th         2
34th         2
40th         2
38th         1
45           1
44           1
Name: count, dtype: int64

In [84]:
df1['floorNum'] = df1['floorNum'].str.extract(r'(-?\d+)')

here we can clearly see that for the given indexes values for all the columns except 'property_name', 'sector', 'society' is not available. so we will remove these indexes because i can't predict values for these many columns

In [85]:
df1[df1['bedRoom'].isnull()]

,property_name,sector,society,price,price_per_sqft,area,areaWithType,bedRoom,bathroom,balcony,additionalRoom,floorNum,facing,agePossession,nearbyLocations,description,furnishDetails,features,rating
2850,2 BHK Flat in Sector 107 Gurgaon,sector 107,signature global solera.,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2851,4 BHK Flat in Sector 53 Gurgaon,sector 53,tulip monsella,NaN,33198.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2852,2 BHK Flat in New Palam Vihar,new palam vihar,my home,NaN,4400.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2853,2 BHK Flat in Sohna,sohna,breez global hill view,NaN,5470.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2923,3 BHK Flat in Sector 99A Gurgaon,sector 99a,pareena coban residences,NaN,5759.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2924,1 BHK Flat in Golf Course Ext Road,golf course ext road,ikon tower baani city centre,NaN,12437.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2925,2 BHK Flat in Sector 89 Gurgaon,sector 89,greenopolis,NaN,4820.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2927,4 BHK Flat in C Block Sushant Lok Phase 1,c block sushant lok phase 1,maple heights,NaN,8888.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2928,2 BHK Flat in Sector 99 Gurgaon,sector 99,assotech blith,NaN,6593.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2994,4 BHK Flat in DLF Phase 4,dlf phase 4,chanderlok society,NaN,7234.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [86]:
df1.drop(index=df1[df1['bedRoom'].isnull()].index, inplace=True)

In [87]:
df1.insert(1, 'property_type', 'flat')

In [88]:
df1.to_csv('flat_cleaned.csv',index=False)